### PR TITLE
Here's the changes we made in webphone.

### PR DESCRIPTION
**Pull request by jimmyxu1985**
jesthtmlreporter
package.json
rcwebphone.code-workspace
src/_test_/ringcentral-web-phone-media-engine.test.js 
are the changes made by Elias for the UT test framework.

changes in session.ts, sipTranportConstructs.ts and index.ts are for the bug fixing.

mediaStream.ts contains the API to collect media stats from peer connection and the API to process reInvite during active call

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -71,9 +71,13 @@
...
     "email": "devsupport@ringcentral.com"
   },
+  "jest": {
+    "testEnvironment": "node",
+    "testResultsProcessor": "./node_modules/jest-html-reporter"
+  },
   "contributors": [
     {
-      "name": "Kirill Konshin"
+      "name": "Kirill Konshin, Elias Sun"
```
Comment (kirill-konshin) [inline package.json]: Should be "contributors": [{"name": "Kirill Konshin"}, {"name": "Elias Sun"}]

File: src/mediaStreams.ts [hunk 2]
```diff
@@ -0,0 +1,402 @@
...
+    } else {
+      console.log(label, msg);
+    }
+  }
+
+  rcWPLogd(label, msg) {
+    if (this.session) {
+      this.session.logger.log(`${label} ${msg}`);
+    } else {
+      console.log(label, msg);
```
Comment (kirill-konshin) [inline src/mediaStreams.ts]: This must be avoided. UA has logger: https://github.com/onsip/SIP.js/blob/master/src/UA.ts#L157

File: src/mediaStreams.ts [hunk 3]
```diff
@@ -0,0 +1,402 @@
...
+      this.session.sessionDescriptionHandler.removeListener('iceConnectionDisconnected', this.onStateChange);
+      this.session.sessionDescriptionHandler.removeListener('iceConnectionClosed', this.onStateChange);
+    }
+  }
+
+  rcWPLoge(label, msg) {
+    if (this.session) {
+      this.session.logger.error(`${label} ${msg}`);
+    } else {
+      console.log(label, msg);
```
Comment (kirill-konshin) [inline src/mediaStreams.ts]: This must be avoided. UA has logger: https://github.com/onsip/SIP.js/blob/master/src/UA.ts#L157

File: jesthtmlreporter.config.json [hunk 4]
```diff
@@ -0,0 +1,5 @@
+{
+    "pageTitle": "RingCentral Webphone Unit Test",
+    "outputPath": "test_report/index.html",
+    "includeFailureMsg": true
+}
```
Comment (kirill-konshin) [inline jesthtmlreporter.config.json]: Jest (with Puppeteer) is already used in this project, is this test addition compatible with the existing tests? Can you make sure they coexist properly?

**Comment by kirill-konshin** [comment]
1. Please fix Lint errors: https://travis-ci.org/ringcentral/ringcentral-web-phone/builds/609089413
2. I've added you as collaborator with write access, please create a branch in main repo and push your commits there, it's required for CI to work
3. Please change `dev` to `master` branch in the PR

@vyshakhbabji please delete `dev` branch to reduce confusion. If you have anything important there — move it to separate branch.

File: src/index.ts [hunk 5]
```diff
@@ -184,6 +184,13 @@ export default class WebPhone {
             allowLegacyNotifications: true
         };
 
+        if (options.builtinEnabled === null || options.builtinEnabled === undefined) {
+            configuration.log.builtinEnabled = true;
+        } else {
+            configuration.log.builtinEnabled = options.builtinEnabled;
+        }
```
Comment (kirill-konshin) [inline src/index.ts]: What was wrong with the original way of passing this down to `log`?

File: src/session.ts [hunk 6]
```diff
@@ -528,10 +531,6 @@ async function warmTransfer(
     target: WebPhoneSession,
     transferOptions: any = {}
 ): Promise<ReferClientContext> {
-    await (this.localHold ? Promise.resolve(null) : this.hold());
-
-    await delay(300);
-
```
Comment (kirill-konshin) [inline src/session.ts]: AFAIR `warmTransfer` requires that call should be on hold. Is it a safe fix? Has anything changed in telco?

File: src/session.ts [hunk 7]
```diff
@@ -565,17 +562,19 @@ async function forward(
...
+                if (this.status === Session.C.STATUS_CONFIRMED) {
+                    clearInterval(interval);
+                    this.mute();
+                    setTimeout(() => {
+                        this.transfer(target, transferOptions);
+                        resolve(null);
+                    }, 700);
+                }
+            }, 50);
+        });
```
Comment (kirill-konshin) [inline src/session.ts]: What is the point of this change (swap `async/await` with `then`)?

File: src/session.ts [hunk 8]
```diff
@@ -551,8 +550,6 @@ async function warmTransfer(
 /*--------------------------------------------------------------------------------------------------------------------*/
 
 async function transfer(this: WebPhoneSession, target: WebPhoneSession, options): Promise<ReferClientContext> {
-    await (this.localHold ? Promise.resolve(null) : this.hold());
-    await delay(300);
```
Comment (kirill-konshin) [inline src/session.ts]: Same here, is it safe to not put the call on hold before blind transfer?

Comment (jimmyxu1985) [reply src/session.ts]: The issue is that hold might not be finished in 300ms, this logic trigger a bug in Jupiter when app send REFER before the reInvite is success, server will return success for the REFER but do nothing. 

In mThor, we won't hold call before we transfer / warmTransfer it.  I also capture the SIP for Spartan and spartan will also NOT hold call before transfer call

File: src/mediaStreams.ts [hunk 9]
```diff
@@ -0,0 +1,402 @@
...
+          return null;
+        }
+    }
+    return null; // no connected peers
+  }
+
+  getRTPReport(reports) {
+    let self = this;
+    let pc = self.session.sessionDescriptionHandler.peerConnection;
+    pc.getStats().then(stats => {
```
[the remark this answers is not in the document]
Comment (jimmyxu1985) [reply src/mediaStreams.ts]: The getStats from github seems have some issue with firefox.

File: jesthtmlreporter.config.json (hunk 4 above)
Comment (jimmyxu1985) [reply jesthtmlreporter.config.json]: The UT framework is actually added by Elias. I'll have a look if we can remove it and use the existed one

Comment (kirill-konshin) [reply jesthtmlreporter.config.json]: No need to remove it. Make them coexist.

**Pull request closed without merging**